File: engine/jvm/java/src/main/java/b1/compiler/Canon.java
package b1.compiler;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

/**
 * B1-CANON-1: strict parsing, canonical serialization and digest.
 * Normative definition: SPEC/20-b1-canon-1.md. Normative implementation: core/c/libb1sig.
 *
 * <p>The JDK ships no JSON parser, so one is written here regardless. That turns out to suit the
 * profile: its guarantees are mostly rejections — duplicate member names, non-integer numbers,
 * non-ASCII keys and unpaired surrogates must all fail loudly — and a permissive parser that
 * silently repaired any of them would produce a digest no other implementation reproduces.
 */
public final class Canon {

    public static final int MAX_DEPTH = 64;
    public static final long MAX_SAFE = 9007199254740991L; // 2^53 - 1
    public static final String ZERO_LINK = "b1c1:" + "0".repeat(64);

    private Canon() {}

    /** Carries a stable B1_ERR_* token, which is what the conform protocol prints. */
    public static class B1Exception extends RuntimeException {
        public final String token;

        public B1Exception(String token, String detail) {
            super(token + ": " + detail);
            this.token = token;
        }
    }

    // --- value model ---------------------------------------------------------
    //
    // There is deliberately no floating-point variant: R1 forbids non-integer numbers, so a
    // document the profile could not canonicalize cannot be represented in the first place.

    public sealed interface Json
            permits JNull, JBool, JInt, JStr, JArr, JObj {}

    public record JNull() implements Json {}

    public record JBool(boolean value) implements Json {}

    public record JInt(long value) implements Json {}

    public record JStr(String value) implements Json {}

    public record JArr(List<Json> items) implements Json {}

    /** TreeMap keeps members in sorted order, which for ASCII-only names (R2) is canonical order. */
    public record JObj(TreeMap<String, Json> members) implements Json {
        public Json get(String key) {
            return members.get(key);
        }
    }

    public static JObj obj() {
        return new JObj(new TreeMap<>());
    }

    public static JObj put(JObj o, String key, Json value) {
        o.members().put(key, value);
        return o;
    }

    public static boolean keySyntaxOk(String k) {
        int n = k.length();
        if (n < 1 || n > 64) return false;
        for (int i = 0; i < n; i++) {
            char c = k.charAt(i);
            boolean ok = (c >= 'A' && c <= 'Z') || (c >= 'a' && c <= 'z') || (c >= '0' && c <= '9')
                    || c == '_' || c == '$' || c == '.' || c == '-';
            if (!ok) return false;
        }
        return true;
    }

    // --- parser --------------------------------------------------------------

    private static final class Parser {
        private final String s;
        private int i;

        Parser(String s) {
            this.s = s;
        }

        void ws() {
            while (i < s.length()) {
                char c = s.charAt(i);
                if (c == ' ' || c == '\t' || c == '\n' || c == '\r') i++;
                else return;
            }
        }

        void lit(String word) {
            if (s.startsWith(word, i)) i += word.length();
            else throw new B1Exception("B1_ERR_PARSE", "expected " + word + " at " + i);
        }

        Json value(int depth) {
            if (depth > MAX_DEPTH) throw new B1Exception("B1_ERR_DEPTH", "depth > " + MAX_DEPTH);
            if (i >= s.length()) throw new B1Exception("B1_ERR_PARSE", "unexpected end of input");
            char c = s.charAt(i);
            if (c == '{') return object(depth);
            if (c == '[') return array(depth);
            if (c == '"') return new JStr(str());
            if (c == 't') { lit("true"); return new JBool(true); }
            if (c == 'f') { lit("false"); return new JBool(false); }
            if (c == 'n') { lit("null"); return new JNull(); }
            if (c == '-' || (c >= '0' && c <= '9')) return new JInt(number());
            throw new B1Exception("B1_ERR_PARSE", "unexpected character at " + i);
        }

        Json object(int depth) {
            i++;
            TreeMap<String, Json> map = new TreeMap<>();
            ws();
            if (i < s.length() && s.charAt(i) == '}') { i++; return new JObj(map); }
            while (true) {
                ws();
                if (i >= s.length() || s.charAt(i) != '"')
                    throw new B1Exception("B1_ERR_PARSE", "expected key at " + i);
                String key = str();
                if (!keySyntaxOk(key)) throw new B1Exception("B1_ERR_KEY_SYNTAX", key);
                if (map.containsKey(key)) throw new B1Exception("B1_ERR_DUPLICATE_KEY", key);
                ws();
                if (i >= s.length() || s.charAt(i) != ':')
                    throw new B1Exception("B1_ERR_PARSE", "expected ':' at " + i);
                i++;
                ws();
                map.put(key, value(depth + 1));
                ws();
                if (i >= s.length()) throw new B1Exception("B1_ERR_PARSE", "unterminated object");
                char c = s.charAt(i);
                if (c == ',') { i++; continue; }
                if (c == '}') { i++; return new JObj(map); }
                throw new B1Exception("B1_ERR_PARSE", "expected ',' or '}' at " + i);
            }
        }

        Json array(int depth) {
            i++;
            List<Json> out = new ArrayList<>();
            ws();
            if (i < s.length() && s.charAt(i) == ']') { i++; return new JArr(out); }
            while (true) {
                ws();
                out.add(value(depth + 1));
                ws();
                if (i >= s.length()) throw new B1Exception("B1_ERR_PARSE", "unterminated array");
                char c = s.charAt(i);
                if (c == ',') { i++; continue; }
                if (c == ']') { i++; return new JArr(out); }
                throw new B1Exception("B1_ERR_PARSE", "expected ',' or ']' at " + i);
            }
        }

        long number() {
            int start = i;
            if (s.charAt(i) == '-') i++;
            int digitsStart = i;
            while (i < s.length() && s.charAt(i) >= '0' && s.charAt(i) <= '9') i++;
            if (i == digitsStart) throw new B1Exception("B1_ERR_PARSE", "expected digits at " + start);
            if (i - digitsStart > 1 && s.charAt(digitsStart) == '0')
                throw new B1Exception("B1_ERR_PARSE", "leading zero at " + digitsStart);
            if (i < s.length()) {
                char c = s.charAt(i);
                if (c == '.' || c == 'e' || c == 'E')
                    throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "at " + start);
            }
            String text = s.substring(start, i);
            if (text.equals("-0")) throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "negative zero");
            long v;
            try {
                v = Long.parseLong(text);
            } catch (NumberFormatException e) {
                throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "out of range: " + text);
            }
            if (v > MAX_SAFE || v < -MAX_SAFE)
                throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "out of range: " + text);
            return v;
        }

        int hex4() {
            if (i + 4 > s.length()) throw new B1Exception("B1_ERR_PARSE", "truncated \\u escape");
            int v = 0;
            for (int k = 0; k < 4; k++) {
                char c = s.charAt(i + k);
                v <<= 4;
                if (c >= '0' && c <= '9') v |= c - '0';
                else if (c >= 'a' && c <= 'f') v |= c - 'a' + 10;
                else if (c >= 'A' && c <= 'F') v |= c - 'A' + 10;
                else throw new B1Exception("B1_ERR_PARSE", "bad \\u escape at " + i);
            }
            i += 4;
            return v;
        }

        /**
         * Decodes one {@code \\uXXXX}, joining a surrogate pair into a single scalar. Java strings
         * are UTF-16, so an unpaired half would survive here and only fail later — it is rejected
         * at the point it is read instead.
         */
        void unicodeEscape(StringBuilder out) {
            int cp = hex4();
            if (cp >= 0xD800 && cp <= 0xDBFF) {
                if (i + 2 > s.length() || s.charAt(i) != '\\' || s.charAt(i + 1) != 'u')
                    throw new B1Exception("B1_ERR_INVALID_UTF8", "unpaired high surrogate");
                i += 2;
                int low = hex4();
                if (low < 0xDC00 || low > 0xDFFF)
                    throw new B1Exception("B1_ERR_INVALID_UTF8", "high surrogate not followed by a low one");
                out.appendCodePoint(0x10000 + ((cp - 0xD800) << 10) + (low - 0xDC00));
                return;
            }
            if (cp >= 0xDC00 && cp <= 0xDFFF)
                throw new B1Exception("B1_ERR_INVALID_UTF8", "unpaired low surrogate");
            out.appendCodePoint(cp);
        }

        String str() {
            i++; // opening quote
            StringBuilder out = new StringBuilder();
            while (true) {
                if (i >= s.length()) throw new B1Exception("B1_ERR_PARSE", "unterminated string");
                char c = s.charAt(i);
                if (c == '"') { i++; return out.toString(); }
                if (c == '\\') {
                    i++;
                    if (i >= s.length()) throw new B1Exception("B1_ERR_PARSE", "unterminated escape");
                    char e = s.charAt(i);
                    i++;
                    switch (e) {
                        case '"' -> out.append('"');
                        case '\\' -> out.append('\\');
                        case '/' -> out.append('/');
                        case 'b' -> out.append('\b');
                        case 'f' -> out.append('\f');
                        case 'n' -> out.append('\n');
                        case 'r' -> out.append('\r');
                        case 't' -> out.append('\t');
                        case 'u' -> unicodeEscape(out);
                        default -> throw new B1Exception("B1_ERR_PARSE", "bad escape \\" + e);
                    }
                    continue;
                }
                if (c < 0x20) throw new B1Exception("B1_ERR_PARSE", "raw control char at " + i);
                out.append(c);
                i++;
            }
        }
    }

    public static Json parse(String text) {
        Parser p = new Parser(text);
        p.ws();
        Json v = p.value(0);
        p.ws();
        if (p.i != text.length()) throw new B1Exception("B1_ERR_PARSE", "trailing input");
        return v;
    }

    // --- serialization -------------------------------------------------------

    private static void escapeInto(StringBuilder out, String s) {
        out.append('"');
        for (int i = 0; i < s.length(); ) {
            int cp = s.codePointAt(i);
            i += Character.charCount(cp);
            switch (cp) {
                case '"' -> out.append("\\\"");
                case '\\' -> out.append("\\\\");
                case '\b' -> out.append("\\b");
                case '\t' -> out.append("\\t");
                case '\n' -> out.append("\\n");
                case '\f' -> out.append("\\f");
                case '\r' -> out.append("\\r");
                default -> {
                    if (cp < 0x20) out.append(String.format("\\u%04x", cp)); // lowercase, per R3
                    else out.appendCodePoint(cp);
                }
            }
        }
        out.append('"');
    }

    private static void canonInto(StringBuilder out, Json v, int depth) {
        if (depth > MAX_DEPTH) throw new B1Exception("B1_ERR_DEPTH", "depth > " + MAX_DEPTH);
        switch (v) {
            case JNull ignored -> out.append("null");
            case JBool b -> out.append(b.value() ? "true" : "false");
            case JInt n -> {
                if (n.value() > MAX_SAFE || n.value() < -MAX_SAFE)
                    throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "out of range");
                out.append(n.value());
            }
            case JStr s -> escapeInto(out, s.value());
            case JArr a -> {
                out.append('[');
                boolean first = true;
                for (Json item : a.items()) {
                    if (!first) out.append(',');
                    first = false;
                    canonInto(out, item, depth + 1);
                }
                out.append(']');
            }
            case JObj o -> {
                out.append('{');
                boolean first = true;
                for (Map.Entry<String, Json> e : o.members().entrySet()) {
                    if (!keySyntaxOk(e.getKey())) throw new B1Exception("B1_ERR_KEY_SYNTAX", e.getKey());
                    if (!first) out.append(',');
                    first = false;
                    escapeInto(out, e.getKey());
                    out.append(':');
                    canonInto(out, e.getValue(), depth + 1);
                }
                out.append('}');
            }
        }
    }

    public static String canonicalize(Json v) {
        StringBuilder out = new StringBuilder();
        canonInto(out, v, 0);
        return out.toString();
    }

    public static String digestValue(Json v) {
        try {
            MessageDigest md = MessageDigest.getInstance("SHA-256");
            byte[] hash = md.digest(canonicalize(v).getBytes(StandardCharsets.UTF_8));
            StringBuilder sb = new StringBuilder(64);
            for (byte b : hash) sb.append(String.format("%02x", b));
            return sb.toString();
        } catch (NoSuchAlgorithmException e) {
            throw new IllegalStateException("SHA-256 unavailable", e);
        }
    }

    public static String digestText(String text) {
        return digestValue(parse(text));
    }

    /** The prefix labels the algorithm and is not part of the hashed input. */
    public static String b1c1(String digestHex) {
        return "b1c1:" + digestHex;
    }
}

File: engine/csharp/B1.Convergence/Canon.cs
using System.Security.Cryptography;
using System.Text;

namespace B1.Convergence;

/// <summary>
/// B1-CANON-1: strict parsing, canonical serialization and digest.
/// Normative definition: SPEC/20-b1-canon-1.md. Normative implementation: core/c/libb1sig.
/// </summary>
/// <remarks>
/// Written independently rather than bound to the C library, so its agreement with the normative
/// implementation is a genuine cross-check. System.Text.Json is not used: it keeps the last of a
/// set of duplicate member names and is permissive about numbers, and both of those silently repair
/// documents this profile is required to reject.
/// </remarks>
public sealed class B1Exception : Exception
{
    public string Token { get; }

    public B1Exception(string token, string detail) : base($"{token}: {detail}") => Token = token;
}

public abstract record Json
{
    public sealed record Null : Json;

    public sealed record Bool(bool Value) : Json;

    public sealed record Int(long Value) : Json;

    public sealed record Str(string Value) : Json;

    public sealed record Arr(IReadOnlyList<Json> Items) : Json;

    /// <summary>SortedDictionary keeps members in canonical order, which for ASCII-only names (R2)
    /// is byte order — so serialization never has to sort and can never forget to.</summary>
    public sealed record Obj(SortedDictionary<string, Json> Members) : Json
    {
        public Json? Get(string key) => Members.TryGetValue(key, out var v) ? v : null;
    }

    public static Obj NewObj() => new(new SortedDictionary<string, Json>(StringComparer.Ordinal));
}

public static class Canon
{
    public const int MaxDepth = 64;
    public const long MaxSafe = 9007199254740991L; // 2^53 - 1

    public static bool KeySyntaxOk(string k)
    {
        if (k.Length is < 1 or > 64) return false;
        foreach (var c in k)
        {
            var ok = c is >= 'A' and <= 'Z' or >= 'a' and <= 'z' or >= '0' and <= '9'
                     or '_' or '$' or '.' or '-';
            if (!ok) return false;
        }
        return true;
    }

    private sealed class Parser(string s)
    {
        private int _i;

        public Json Parse()
        {
            SkipWs();
            var v = Value(0);
            SkipWs();
            if (_i != s.Length) throw new B1Exception("B1_ERR_PARSE", "trailing input");
            return v;
        }

        private void SkipWs()
        {
            while (_i < s.Length && (s[_i] == ' ' || s[_i] == '\t' || s[_i] == '\n' || s[_i] == '\r'))
                _i++;
        }

        private void Literal(string word)
        {
            if (_i + word.Length <= s.Length && string.CompareOrdinal(s, _i, word, 0, word.Length) == 0)
                _i += word.Length;
            else throw new B1Exception("B1_ERR_PARSE", $"expected {word} at {_i}");
        }

        private Json Value(int depth)
        {
            if (depth > MaxDepth) throw new B1Exception("B1_ERR_DEPTH", $"depth > {MaxDepth}");
            if (_i >= s.Length) throw new B1Exception("B1_ERR_PARSE", "unexpected end of input");

            var c = s[_i];
            if (c == '{') return Object(depth);
            if (c == '[') return Array(depth);
            if (c == '"') return new Json.Str(String());
            if (c == 't') { Literal("true"); return new Json.Bool(true); }
            if (c == 'f') { Literal("false"); return new Json.Bool(false); }
            if (c == 'n') { Literal("null"); return new Json.Null(); }
            if (c == '-' || c is >= '0' and <= '9') return new Json.Int(Number());
            throw new B1Exception("B1_ERR_PARSE", $"unexpected character at {_i}");
        }

        private Json Object(int depth)
        {
            _i++;
            var map = new SortedDictionary<string, Json>(StringComparer.Ordinal);
            SkipWs();
            if (_i < s.Length && s[_i] == '}') { _i++; return new Json.Obj(map); }

            while (true)
            {
                SkipWs();
                if (_i >= s.Length || s[_i] != '"') throw new B1Exception("B1_ERR_PARSE", "expected key");
                var key = String();
                if (!KeySyntaxOk(key)) throw new B1Exception("B1_ERR_KEY_SYNTAX", key);
                if (map.ContainsKey(key)) throw new B1Exception("B1_ERR_DUPLICATE_KEY", key);
                SkipWs();
                if (_i >= s.Length || s[_i] != ':') throw new B1Exception("B1_ERR_PARSE", "expected ':'");
                _i++;
                SkipWs();
                map[key] = Value(depth + 1);
                SkipWs();
                if (_i >= s.Length) throw new B1Exception("B1_ERR_PARSE", "unterminated object");
                if (s[_i] == ',') { _i++; continue; }
                if (s[_i] == '}') { _i++; return new Json.Obj(map); }
                throw new B1Exception("B1_ERR_PARSE", "expected ',' or '}'");
            }
        }

        private Json Array(int depth)
        {
            _i++;
            var items = new List<Json>();
            SkipWs();
            if (_i < s.Length && s[_i] == ']') { _i++; return new Json.Arr(items); }

            while (true)
            {
                SkipWs();
                items.Add(Value(depth + 1));
                SkipWs();
                if (_i >= s.Length) throw new B1Exception("B1_ERR_PARSE", "unterminated array");
                if (s[_i] == ',') { _i++; continue; }
                if (s[_i] == ']') { _i++; return new Json.Arr(items); }
                throw new B1Exception("B1_ERR_PARSE", "expected ',' or ']'");
            }
        }

        private long Number()
        {
            var start = _i;
            if (s[_i] == '-') _i++;
            var digitsStart = _i;
            while (_i < s.Length && s[_i] is >= '0' and <= '9') _i++;
            if (_i == digitsStart) throw new B1Exception("B1_ERR_PARSE", "expected digits");
            if (_i - digitsStart > 1 && s[digitsStart] == '0')
                throw new B1Exception("B1_ERR_PARSE", "leading zero");
            if (_i < s.Length && (s[_i] == '.' || s[_i] == 'e' || s[_i] == 'E'))
                throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "non-integer");

            var text = s[start.._i];
            if (text == "-0") throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "negative zero");
            if (!long.TryParse(text, out var v) || v > MaxSafe || v < -MaxSafe)
                throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", $"out of range: {text}");
            return v;
        }

        private int Hex4()
        {
            if (_i + 4 > s.Length) throw new B1Exception("B1_ERR_PARSE", "truncated escape");
            var v = 0;
            for (var k = 0; k < 4; k++)
            {
                var c = s[_i + k];
                v <<= 4;
                v |= c switch
                {
                    >= '0' and <= '9' => c - '0',
                    >= 'a' and <= 'f' => c - 'a' + 10,
                    >= 'A' and <= 'F' => c - 'A' + 10,
                    _ => throw new B1Exception("B1_ERR_PARSE", "bad hex digit")
                };
            }
            _i += 4;
            return v;
        }

        /// <summary>
        /// Decodes one \uXXXX, joining a surrogate pair into a single scalar. .NET strings are
        /// UTF-16, so an unpaired half would survive here and fail only later; it is rejected at
        /// the point it is read.
        /// </summary>
        private void UnicodeEscape(StringBuilder into)
        {
            var cp = Hex4();
            if (cp is >= 0xD800 and <= 0xDBFF)
            {
                if (_i + 2 > s.Length || s[_i] != '\\' || s[_i + 1] != 'u')
                    throw new B1Exception("B1_ERR_INVALID_UTF8", "unpaired high surrogate");
                _i += 2;
                var low = Hex4();
                if (low is < 0xDC00 or > 0xDFFF)
                    throw new B1Exception("B1_ERR_INVALID_UTF8", "high surrogate without a low one");
                into.Append(char.ConvertFromUtf32(0x10000 + ((cp - 0xD800) << 10) + (low - 0xDC00)));
                return;
            }
            if (cp is >= 0xDC00 and <= 0xDFFF)
                throw new B1Exception("B1_ERR_INVALID_UTF8", "unpaired low surrogate");
            into.Append((char)cp);
        }

        private string String()
        {
            _i++; // opening quote
            var b = new StringBuilder();
            while (true)
            {
                if (_i >= s.Length) throw new B1Exception("B1_ERR_PARSE", "unterminated string");
                var c = s[_i];
                if (c == '"') { _i++; return b.ToString(); }
                if (c == '\\')
                {
                    _i++;
                    if (_i >= s.Length) throw new B1Exception("B1_ERR_PARSE", "unterminated escape");
                    var e = s[_i];
                    _i++;
                    switch (e)
                    {
                        case '"': b.Append('"'); break;
                        case '\\': b.Append('\\'); break;
                        case '/': b.Append('/'); break;
                        case 'b': b.Append('\b'); break;
                        case 'f': b.Append('\f'); break;
                        case 'n': b.Append('\n'); break;
                        case 'r': b.Append('\r'); break;
                        case 't': b.Append('\t'); break;
                        case 'u': UnicodeEscape(b); break;
                        default: throw new B1Exception("B1_ERR_PARSE", $"bad escape \\{e}");
                    }
                    continue;
                }
                if (c < 0x20) throw new B1Exception("B1_ERR_PARSE", "raw control character");
                b.Append(c);
                _i++;
            }
        }
    }

    public static Json Parse(string text) => new Parser(text).Parse();

    private static void EscapeInto(StringBuilder o, string s)
    {
        o.Append('"');
        foreach (var ch in s)
        {
            switch (ch)
            {
                case '"': o.Append("\\\""); break;
                case '\\': o.Append("\\\\"); break;
                case '\b': o.Append("\\b"); break;
                case '\t': o.Append("\\t"); break;
                case '\n': o.Append("\\n"); break;
                case '\f': o.Append("\\f"); break;
                case '\r': o.Append("\\r"); break;
                default:
                    if (ch < 0x20) o.Append("\\u").Append(((int)ch).ToString("x4")); // lowercase
                    else o.Append(ch);
                    break;
            }
        }
        o.Append('"');
    }

    private static void CanonInto(StringBuilder o, Json v, int depth)
    {
        if (depth > MaxDepth) throw new B1Exception("B1_ERR_DEPTH", $"depth > {MaxDepth}");
        switch (v)
        {
            case Json.Null:
                o.Append("null");
                break;
            case Json.Bool b:
                o.Append(b.Value ? "true" : "false");
                break;
            case Json.Int n:
                if (n.Value > MaxSafe || n.Value < -MaxSafe)
                    throw new B1Exception("B1_ERR_NONINTEGER_NUMBER", "out of range");
                o.Append(n.Value);
                break;
            case Json.Str s:
                EscapeInto(o, s.Value);
                break;
            case Json.Arr a:
                o.Append('[');
                for (var i = 0; i < a.Items.Count; i++)
                {
                    if (i > 0) o.Append(',');
                    CanonInto(o, a.Items[i], depth + 1);
                }
                o.Append(']');
                break;
            case Json.Obj obj:
                o.Append('{');
                var first = true;
                foreach (var (k, val) in obj.Members)
                {
                    if (!KeySyntaxOk(k)) throw new B1Exception("B1_ERR_KEY_SYNTAX", k);
                    if (!first) o.Append(',');
                    first = false;
                    EscapeInto(o, k);
                    o.Append(':');
                    CanonInto(o, val, depth + 1);
                }
                o.Append('}');
                break;
            default:
                throw new B1Exception("B1_ERR_PARSE", "unknown node");
        }
    }

    public static string Canonicalize(Json v)
    {
        var b = new StringBuilder();
        CanonInto(b, v, 0);
        return b.ToString();
    }

    public static string DigestValue(Json v)
    {
        var bytes = SHA256.HashData(Encoding.UTF8.GetBytes(Canonicalize(v)));
        // Lowercase is normative (SPEC/20 §3). ToHexString emits uppercase and the lowercase
        // overload is .NET 9+, so the case is forced here rather than assumed.
        return Convert.ToHexString(bytes).ToLowerInvariant();
    }

    public static string DigestText(string text) => DigestValue(Parse(text));
}
